### ISSUE TITLE
Pytorch 0.4.0: Model behavior changes heavily after save and load weights

### ISSUE THREAD

**Issue by saturdays** [pytorch/pytorch#8820]
# Issue description
During training I save model by using: torch.save(model.state_dict(), file). And I reach 99% accuracy on both test and train data set.
Then, I load the model and test again using: model.load_state_dict(checkpoint, strict=True). The accuracy drops to 0.1%. An initial value!!

My net is quite simple, so I print all the weights in the checkpoint both before and after validation. They are exactly the same.

I thought the drop_out layer causes this issue, but remove it doesn't work either.

```python
## Code example
To save:
    model = Net()
    model = Net().to(device)
    optimizer = torch.optim.SGD(model.parameters(), lr=args.lr,
                                momentum=args.momentum,
                                weight_decay=1e-5)
    while acc < 0.98:
        epoch += 1
        adjust_learning_rate(optimizer, epoch)
        acc1 = train(args, model, device, train_loader, optimizer, epoch)
        acc2 = test(args, model, device, test_loader)
        acc = (acc1+acc2)*0.5
        torch.save(model.state_dict(), 'checkpoints/mnist_gpu_%d.pkl'%epoch)

To load:
    model = Net()
    model.eval()
    model = Net().to(device)
    checkpoint = torch.load('checkpoints/mnist_gpu_20.pkl')
    for k,v in enumerate(checkpoint):
        np.savetxt('%s_pre'%v, checkpoint[v].cpu().numpy().reshape(-1))
    model.load_state_dict(checkpoint, strict=True)
    acc2 = test(args, model, device, test_loader)
    checkpoint = model.state_dict()
    for k,v in enumerate(checkpoint):
        np.savetxt('%s_after'%v, checkpoint[v].cpu().numpy().reshape(-1))

#model is super simple:
import torch
import torch.nn as nn
import torch.nn.functional as F
class Fire(torch.nn.Module):
    def __init__(self,inchn,sqzout_chn,exp1x1out_chn,exp3x3out_chn):
        super(Fire,self).__init__()
        self.inchn = inchn
        self.squeeze = torch.nn.Conv2d(inchn,sqzout_chn,kernel_size=1)
        self.squeeze_act = torch.nn.ReLU(inplace=True)
        self.expand1x1 = torch.nn.Conv2d(sqzout_chn,exp1x1out_chn,kernel_size=1)
        self.expand1x1_act = torch.nn.ReLU(inplace=True)
        self.expand3x3 = torch.nn.Conv2d(sqzout_chn,exp3x3out_chn,kernel_size=3, padding=1)
        self.expand3x3_act = torch.nn.ReLU(inplace=True)
    def forward(self, x):
        x = self.squeeze_act(self.squeeze(x))
        return torch.cat([
                self.expand1x1_act(self.expand1x1(x)),
                self.expand3x3_act(self.expand3x3(x))
                ], 1)
class Net(nn.Module):
    def __init__(self, num_class=10):
        super(Net, self).__init__()
        self.num_class = 10
        self.features = torch.nn.Sequential(
            torch.nn.Conv2d(3,96,kernel_size=3),
            torch.nn.ReLU(inplace=True),
            torch.nn.MaxPool2d(kernel_size=2, ceil_mode=False),
            Fire(96,16,64,64),
            torch.nn.MaxPool2d(kernel_size=2, ceil_mode=False),
            Fire(128,32,128,128),
            torch.nn.MaxPool2d(kernel_size=2, ceil_mode=False),
            Fire(256,64,256,256),
        )
        final_conv = torch.nn.Conv2d(512,self.num_class,kernel_size=1)
        self.classifier = torch.nn.Sequential(
            final_conv,
            torch.nn.ReLU(inplace=True),
            torch.nn.AdaptiveAvgPool2d(1)
        )
        for m in self.modules():
            if isinstance(m, torch.nn.Conv2d):
                if m is final_conv:
                    torch.nn.init.normal_(m.weight.data, mean=0.0, std=0.1)
                else:
                    torch.nn.init.kaiming_uniform_(m.weight.data)
                if m.bias is not None:
                    m.bias.data.zero_()
    def forward(self, x):
        x = self.features(x)
        x = self.classifier(x)
        x = x.view(-1, self.num_class)
        return x
```

## System Info
Please copy and paste the output from our
Collecting environment information...
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Ubuntu 16.04.3 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 8.0.44
GPU models and configuration:
GPU 0: TITAN X (Pascal)
GPU 1: TITAN X (Pascal)
GPU 2: TITAN X (Pascal)
GPU 3: TITAN X (Pascal)

Nvidia driver version: 381.09
cuDNN version: Probably one of the following:
/usr/local/cuda-8.0/lib64/libcudnn.so
/usr/local/cuda-8.0/lib64/libcudnn.so.5
/usr/local/cuda-8.0/lib64/libcudnn.so.5.0.5
/usr/local/cuda-8.0/lib64/libcudnn.so.5.1.10
/usr/local/cuda-8.0/lib64/libcudnn.so.6
/usr/local/cuda-8.0/lib64/libcudnn.so.6.0.21
/usr/local/cuda-8.0/lib64/libcudnn_static.a
/usr/local/lib/libcudnn.so.7.0.5
/usr/local/lib/libcudnn_static.a
/usr/local/lib/python2.7/dist-packages/torch/lib/libcudnn-a2b758a6.so.7.0.3
/usr/local/lib/python3.5/dist-packages/torch/lib/libcudnn-900fef33.so.7.0.5

Versions of relevant libraries:
[pip3] numpy (1.13.3)
[pip3] numpydoc (0.7.0)
[pip3] torch (0.4.0)
[pip3] torchvision (0.2.1)
[conda] pytorch                   0.4.0           py36_cuda8.0.61_cudnn7.1.2_1    pytorch
[conda] torch                     0.4.0                     <pip>
[conda] torchvision               0.2.1                    py36_1    pytorch

**Comment by saturdays** [comment]
Here is another test:
(1) I save and load the weights every epoch, the model works fine and converges during training:
epoch=1
while acc < 0.97:
     epoch += 1
     acc1 = train(args, model, device, train_loader, optimizer, epoch)
     torch.save(model.state_dict(), 'checkpoints/mnist_gpu_%d.pkl'%epoch)
     model.load_state_dict(torch.load('checkpoints/mnist_gpu_%d.pkl'%epoch), strict=True)\
     acc2 = test(args, model, device, test_loader)
     acc = (acc1+acc2)*0.5
(2) But, If a comment train and only do inference. Model changed. Acc goes to 1%
   epoch = 69
    while acc < 0.97:
        epoch += 1
        adjust_learning_rate(optimizer, epoch)
        #acc1 = train(args, model, device, train_loader, optimizer, epoch)
        #torch.save(model.state_dict(), 'checkpoints/mnist_gpu_%d.pkl'%epoch)
        model.load_state_dict(torch.load('checkpoints/mnist_gpu_%d.pkl'%epoch), strict=True)
        acc2 = test(args, model, device, test_loader)
        print(acc2)
        break

**Comment by saturdays** [comment]
Here is another test, I run the inference script for three times, without change anything. The result is different everytime
python validate.py
0.1375
python validate.py
0.0025
python validate.py
0.091

**Comment by saturdays** [comment]
Here is another test that much weird.
Instead of inference in a separate script, I run validation immediately after train by save and reload. The inference result is correct. Code is:

    epoch =0
    while epoch<10:
        epoch += 1
        adjust_learning_rate(optimizer, epoch)
        acc1 = train(args, model, device, train_loader, optimizer, epoch)
        torch.save(model, 'checkpoints/mnist_gpu_%d.pt'%epoch)
        acc2 = test(args, model, device, test_loader)
        acc = (acc1+acc2)*0.5
    model = torch.load('checkpoints/mnist_gpu_%d.pt'%epoch)
    acc2 = test(args, model, device, test_loader)
    print(acc2)

**Comment by saturdays** [comment]
Same code works fine on Python2.7.  but not on Python3.6.

**Comment by SsnL** [comment]
That indeed looks rather weird! Could you share a full runnable script(s) with me so I can take a look?

**Comment by saturdays** [comment]
Of course. But runnable script depends on a lot of img data. How can I share them with you?

**Comment by SsnL** [comment]
@saturdays Can you try to adapt the model to some standard dataset, or a synthetically generated one?

**Comment by SsnL** [comment]
Or a trained checkpoint file, and ~50 images that we can test the model on.

**Comment by saturdays** [comment]
OK. I can do this. But How can I give you the data, this page doesn't support attachment?

**Comment by SsnL** [comment]
I see that you are training on MNIST, so I suppose just the code should be fine?

**Comment by saturdays** [comment]
no, I'm training on a color digits data set. Problem is similar to MNIST but different net structure.

**Comment by SsnL** [comment]
Okay. If you can, please send the code & file to ssnl@fb.com. Thanks!

**Comment by saturdays** [comment]
Sent to u. Please check. Thanks a lot.

**Comment by saturdays** [comment]
Hi, do we have a fix?

**Comment by SsnL** [comment]
Hi, in your `mydataset.py` (full script [here](https://gist.github.com/SsnL/84dea832cebf4ccc44c6f7126352508f)), there is this part
```py
        for i, lab in enumerate(set(self.label_nms)):
            self.label_ids[lab] = i
```

Since `set` is unordered, the mapping between `label_nm` and `label_id` is not deterministic. (They maybe are Py2, but that is undefined behavior.) So you see different results on train vs test, and different test runs. 

Since those in `self.label_nms` are just strings of digits, I changed the dataset to just store `int(label_nm)` as labels. Train + test works as expected. [Here](https://gist.github.com/SsnL/5946f101e2ed7e83292054a1af258632) is a modified dataset file.

```
Test set: Average loss: 1.7855, Accuracy: 44/90 (49%)
Train Epoch: 20 [0/90 (0%)]     Loss: 1.765595  Accuuracy: 0.48
Test set: Average loss: 1.7540, Accuracy: 43/90 (48%)
Test set: Average loss: 1.7540, Accuracy: 43/90 (48%)
acc2= 0.4777777777777778
➜  ld_weights_debug python validate.py
0.4777777777777778
➜  ld_weights_debug
```
